File: Ch01/Ex1_11.hs
{-
 - Ex1_11.hs
 - Kevin C. Baird
 - SICP in Haskell
 -
 - Define both recursive and iterative processes for f
 - (Exercise 1.11)
 -
 -}

module Ex1_11 where

fRec :: Int -> Int
fRec n
  | n < 3     = n
  | otherwise = (fRec (n-1)) +
    (fRec (n-2) * 2) +
    (fRec (n-3) * 3)

fIter :: Int -> Int
fIter n = fIter2 n 0

fIter2 :: Int -> Int -> Int
fIter2 n total
  | n < 3     = n + total
  | otherwise = (fIter2 (n-1) total) +
    ((fIter2 (n-2) total) * 2) +
    ((fIter2 (n-3) total) * 3)

File: Ch01/Ex1_11.scala
/**
  * Ex1_11.scala
  * Kevin C. Baird
  * SICP in Scala (http://www.scala-lang.org/documentation/)
  *
  * Define both recursive and iterative processes for f
  * (Exercise 1.11)
  */

def fRec(n: Int): Int = {
  (n < 3) match {
    case true  => n
    case false => fRec(n-1) + (fRec(n-2)) * 2 + (fRec(n-3)) * 3
  }
}

def fIter(n: Int) = {
  def fInnerIter(n: Int, total: Int): Int = {
    (n < 3) match {
      case true => n + total
      case false => fInnerIter((n-1), total) +
        (fInnerIter((n-2), total) * 2) +
        (fInnerIter((n-3), total) * 3)
    }
  }
  fInnerIter(n, 0)
}

/**
  * execute with
  * scala -i Ex1_11.scala -e "println(fRec(4))"
  * scala -i Ex1_11.scala -e "println(fIter(4))"
  * until I figure out a better way
  */

